xonsh/jupyter_kernel.py

```diff
@@ -1,72 +1,398 @@
+        )
+        self.stdin_stream = zmqstream.ZMQStream(self.stdin_socket)
+        self.stdin_stream.on_recv(self.stdin_handler)
+
+        # Shell
+        self.shell_socket = ctx.socket(zmq.ROUTER)
+        self.config["shell_port"] = bind(
+            self.shell_socket, connection, self.config["shell_port"]
+        )
+        self.shell_stream = zmqstream.ZMQStream(self.shell_socket)
+        self.shell_stream.on_recv(self.shell_handler)
+
+        # start up configurtation
+        self.dprint(2, "Config:", json.dumps(self.config))
+        self.dprint(1, "Starting loops...")
+        self.hb_thread = threading.Thread(target=self.heartbeat_loop)
+        self.hb_thread.daemon = True
+        self.hb_thread.start()
+        self.dprint(1, "Ready! Listening...")
+        ioloop.IOLoop.instance().start()
+
+    def shutdown(self):
+        """Shutsdown the kernel"""
+        self.exiting = True
+        ioloop.IOLoop.instance().stop()
+
+    def dprint(self, level, *args, **kwargs):
+        """Print but with debug information."""
+        if level <= self.debug_level:
+            print("DEBUG" + str(level) + ":", file=sys.__stdout__, *args, **kwargs)
+            sys.__stdout__.flush()
+
+    def sign(self, messages):
+        """Sign a message list with a secure signature."""
+        h = self.auth.copy()
+        for m in messages:
+            h.update(m)
+        return h.hexdigest().encode("ascii")
+
+    def new_header(self, message_type):
+        """Make a new header"""
+        return {
+            "date": datetime.datetime.now().isoformat(),
+            "msg_id": str(uuid.uuid4()),
+            "username": "kernel",
+            "session": self.session_id,
+            "msg_type": message_type,
+            "version": "5.0",
+        }
+
+    def send(
+        self,
+        stream,
+        message_type,
+        content=None,
+        parent_header=None,
+        metadata=None,
+        identities=None,
+    ):
+        """Send data to the client via a stream"""
+        header = self.new_header(message_type)
+        if content is None:
+            content = {}
+        if parent_header is None:
+            parent_header = {}
+        if metadata is None:
+            metadata = {}
+
+        messages = list(map(dump_bytes, [header, parent_header, metadata, content]))
+        signature = self.sign(messages)
+        parts = [DELIM, signature] + messages
+        if identities:
+            parts = identities + parts
+        self.dprint(3, "send parts:", parts)
+        stream.send_multipart(parts)
+        if isinstance(stream, zmqstream.ZMQStream):
+            stream.flush()
+
+    def deserialize_wire_message(self, wire_message):
+        """Split the routing prefix and message frames from a message on the wire"""
+        delim_idx = wire_message.index(DELIM)
+        identities = wire_message[:delim_idx]
+        m_signature = wire_message[delim_idx + 1]
+        msg_frames = wire_message[delim_idx + 2 :]
```

Comment:
pep8 space before colon

Reply:
Unfortunately, this is technically [valid PEP8](https://www.python.org/dev/peps/pep-0008/#pet-peeves), 

> However, in a slice the colon acts like a binary operator, and should have equal amounts on either side (treating it as the operator with the lowest priority)

I just used `black` to reformat the file.  flake8 incorrectly says that this is an error :(

Reply:
That's obnoxious.

Reply:
Yeah it is pretty much the only part of black I don't like... But since I don't actually have to write code that way , I don't think I mind it :)